torch/csrc/jit/tracer.h

```diff
@@ -216,34 +213,35 @@ void addInputs(Node *n, const char * name, std::array<bool, N> value) {
   throw std::runtime_error("Found an unsupported argument type in the JIT tracer. File a bug report.");
 }
 
-template<typename T, typename... Args>
-void addInputs(Node *n, const char * arg_name, T arg, const char * next_arg_name, Args... args) {
-  addInputs(n, arg_name, arg);
-  addInputs(n, next_arg_name, args...);
-}
-
-} // namespace detail
+TORCH_API void postRecordTrace(Node* node, at::ArrayRef<Variable> outputs);
 
-// NB: if you change this function, you might want to take a look at
-// preRecordPythonTrace from python_tracer.cpp
-template<typename... Args>
-PreTraceInfo preRecordTrace(Symbol op, Args... inputs) {
-  PreTraceInfo info;
-  auto & state = getTracingState();
-  auto & graph = state->graph;
-
-  Node * n = info.n = graph->create(op, /*outputs=*/0);
-  recordSourceLocation(n);
-
-  detail::addInputs(n, inputs...);
+// All these overloads are to handle (1) single `at::Tensor` (2)
+// `vector<at::Tensor>` because of implicit conversion trouble with `Variable`,
+// as well as any other types, for which we throw an exception.
 
-  // NB: Order matters. This must append after inputs but before outputs.
-  graph->appendNode(n);
+inline void postRecordTrace(
+    Node* node,
+    at::ArrayRef<at::Tensor> tensors) {
+  postRecordTrace(node, fmap(tensors, [](const at::Tensor& tensor) {
+                    return Variable(tensor);
+                  }));
+}
 
-  return info;
+inline void postRecordTrace(Node* node, at::Tensor tensor) {
+  postRecordTrace(node, at::ArrayRef<Variable>{Variable(tensor)});
 }
 
-TORCH_API void postRecordTrace(const PreTraceInfo& info, at::ArrayRef<Variable> outputs);
+template <
+    typename T,
+    typename = torch::enable_if_t<
+        (!std::is_convertible<torch::decay_t<T>, ArrayRef<Variable>>::value &&
+         !std::is_convertible<torch::decay_t<T>, ArrayRef<at::Tensor>>::value &&
+         !std::is_convertible<torch::decay_t<T>, Variable>::value)>>
```

Comment:
That doesn't look like getting rid of metaprogramming to me 😄 Wouldn't it be better to `= delete` this function to get a compile time error signal?

Reply:
Haha yeah you're right about the MP. I think with Zach we agreed that the runtime error would be much less cryptic to users than a compile time error for a deleted function